File: composeApp/src/commonTest/kotlin/com/aus/notelikeus/domain/query/SearchTextTest.kt
package com.aus.notelikeus.domain.query

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue

class SearchTextTest {

    @Test
    fun `folds case`() {
        assertEquals("hello world", foldForSearch("Hello WORLD"))
    }

    @Test
    fun `strips diacritics so an accented note is findable by plain letters`() {
        // The motivating case: `contains(ignoreCase = true)` folds case and nothing else, so
        // searching "cafe" never found "café".
        assertEquals("cafe", foldForSearch("café"))
        assertEquals("uber", foldForSearch("Über"))
        assertEquals("naive", foldForSearch("naïve"))
        assertEquals("zurich", foldForSearch("Zürich"))
        assertEquals("garcon", foldForSearch("garçon"))
        assertEquals("smorgasbord", foldForSearch("smörgåsbord"))
    }

    @Test
    fun `expands the ligatures and eszett rather than dropping them`() {
        assertEquals("straße".let { foldForSearch(it) }, "strasse")
        assertEquals("aether", foldForSearch("æther"))
        assertEquals("oeuvre", foldForSearch("œuvre"))
    }

    @Test
    fun `covers Latin Extended-A`() {
        assertEquals("lodz", foldForSearch("Łódź"))
        assertEquals("cesky", foldForSearch("Český"))
        assertEquals("gdansk", foldForSearch("Gdańsk"))
        assertEquals("istanbul", foldForSearch("İstanbul"))
    }

    @Test
    fun `punctuation separates rather than glues`() {
        // "don't" -> don, t. Gluing them would invent "dont", a word the note does not contain.
        assertEquals("don t", foldForSearch("don't"))
        assertEquals("a b", foldForSearch("a---b"))
        assertEquals("hello world", foldForSearch("  hello,   world!  "))
    }

    @Test
    fun `keeps digits`() {
        assertEquals("meeting 2026 q3", foldForSearch("Meeting 2026 Q3"))
    }

    @Test
    fun `passes unfolded scripts through instead of deleting them`() {
        // No folding table for these, but they must stay searchable by their own characters.
        // Dropping them would make those notes unfindable, which is strictly worse.
        assertEquals("привет", foldForSearch("Привет"))
        assertEquals("日本語", foldForSearch("日本語"))
    }

    @Test
    fun `empty and separator-only input fold to empty`() {
        assertEquals("", foldForSearch(""))
        assertEquals("", foldForSearch("   "))
        assertEquals("", foldForSearch("--- ,,, ---"))
    }

    @Test
    fun `tokenises in order and keeps duplicates`() {
        assertEquals(listOf("the", "cat", "the", "hat"), searchTokens("The cat the HAT"))
        assertEquals(emptyList(), searchTokens("!!!"))
    }

    @Test
    fun `search text covers every field the old in-memory search looked at`() {
        val text = buildSearchText(
            title = "Réunion",
            content = "Discuss the björk release",
            checklistTexts = listOf("Buy tickets", "Café after"),
            labelNames = listOf("Musique")
        )
        listOf("reunion", "bjork", "tickets", "cafe", "musique").forEach {
            assertTrue(it in text, "'$it' missing from '$text'")
        }
    }

    @Test
    fun `search text skips blank parts without leaving double spaces`() {
        val text = buildSearchText("Title", "", emptyList(), listOf("", "Work"))
        assertEquals("title work", text)
    }

    @Test
    fun `folding is idempotent`() {
        // The column is rebuilt on every write; folding an already-folded value must not drift.
        val once = buildSearchText("Café Zürich", "Łódź", listOf("naïve"), listOf("Œuvre"))
        assertEquals(once, foldForSearch(once))
    }
}

File: web/src/lib/text/searchText.test.ts
import { describe, expect, it } from 'vitest';
import {
  buildSearchText,
  foldForSearch,
  noteMatchesSearchQuery,
  searchTokens,
} from '@/lib/text/searchText';

describe('foldForSearch', () => {
  it('folds case', () => {
    expect(foldForSearch('Hello WORLD')).toBe('hello world');
  });

  it('strips diacritics so an accented note is findable by plain letters', () => {
    expect(foldForSearch('café')).toBe('cafe');
    expect(foldForSearch('Über')).toBe('uber');
    expect(foldForSearch('naïve')).toBe('naive');
    expect(foldForSearch('Zürich')).toBe('zurich');
    expect(foldForSearch('garçon')).toBe('garcon');
    expect(foldForSearch('smörgåsbord')).toBe('smorgasbord');
  });

  it('expands ligatures and eszett rather than dropping them', () => {
    expect(foldForSearch('straße')).toBe('strasse');
    expect(foldForSearch('æther')).toBe('aether');
    expect(foldForSearch('œuvre')).toBe('oeuvre');
  });

  it('covers Latin Extended-A', () => {
    expect(foldForSearch('Łódź')).toBe('lodz');
    expect(foldForSearch('Český')).toBe('cesky');
    expect(foldForSearch('Gdańsk')).toBe('gdansk');
    expect(foldForSearch('İstanbul')).toBe('istanbul');
  });

  it('punctuation separates rather than glues', () => {
    expect(foldForSearch("don't")).toBe('don t');
    expect(foldForSearch('a---b')).toBe('a b');
    expect(foldForSearch('  hello,   world!  ')).toBe('hello world');
  });

  it('keeps digits and passes unfolded scripts through', () => {
    expect(foldForSearch('Meeting 2026 Q3')).toBe('meeting 2026 q3');
    expect(foldForSearch('Привет')).toBe('привет');
    expect(foldForSearch('日本語')).toBe('日本語');
  });

  it('tokenises in order and keeps duplicates', () => {
    expect(searchTokens('The cat the HAT')).toEqual(['the', 'cat', 'the', 'hat']);
    expect(searchTokens('!!!')).toEqual([]);
  });

  it('search text covers every field the in-memory filter looks at', () => {
    const text = buildSearchText(
      'Réunion',
      'Discuss the björk release',
      ['Buy tickets', 'Café after'],
      ['Musique'],
    );
    for (const token of ['reunion', 'bjork', 'tickets', 'cafe', 'musique']) {
      expect(text).toContain(token);
    }
  });
});

describe('noteMatchesSearchQuery', () => {
  it('matches by word prefix, not substring', () => {
    expect(noteMatchesSearchQuery('notes', 'not')).toBe(true);
    expect(noteMatchesSearchQuery('notes', 'ote')).toBe(false);
  });

  it('requires every token and ignores order', () => {
    expect(noteMatchesSearchQuery('bread and milk', 'milk bread')).toBe(true);
    expect(noteMatchesSearchQuery('bread and milk', 'milk cheese')).toBe(false);
  });

  it('treats blank queries as a match', () => {
    expect(noteMatchesSearchQuery('anything', '   ')).toBe(true);
  });
});
